### PR TITLE
Add `Interests` for registering process(Issue #908)

### PR DESCRIPTION
**Pull request by FXTi**
Add new `Interests` based on mio-st version and `Ready` struct.

### CODE REVIEW

**Comment by carllerche** [comment]
@FXTi Thanks for doing this. Could you merge master? This should fix CI.

@Thomasdezeeuw does this look good to you?

File: src/event_imp.rs [hunk 1]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
     assert_eq!("Writable", format!("{:?}", Ready::writable()));
 }
 
+
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Double empty line.

File: src/event_imp.rs [hunk 2]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
...
+/// writable, no event will be returned from [`poll`].
+///
+/// [`Poll`]: struct.Poll.html                                                 
+/// [`readable`]: #method.readable
+/// [`writable`]: #method.writable
+#[derive(Copy, PartialEq, Eq, Clone, PartialOrd, Ord)]
+pub struct Interests(usize);
+
+impl Interests {
+
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Empty line.

File: src/event_imp.rs [hunk 3]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
...
+    /// Returns a `Interests` representing readable readiness.
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Remove "a".

File: src/event_imp.rs [hunk 4]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
...
+    /// assert!(interest.is_readable());
+    /// ```
+    ///
+    /// [`Poll`]: struct.Poll.html
+    #[inline]
+    pub fn readable() -> Interests {
+        Interests(READABLE)
+    }
+
+    /// Returns a `Interests` representing writable readiness.
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Same as above.

File: src/event_imp.rs [hunk 5]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
...
+    /// assert!(interest.is_writable());
+    /// ```
+    ///
+    /// [`Poll`]: struct.Poll.html
+    #[inline]
+    pub fn writable() -> Interests {
+        Interests(WRITABLE)
+    }
+
+    /// Returns a `Interests` representing readable and writable readiness.
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: And again remove "a".

File: src/event_imp.rs [hunk 6]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
...
+    /// assert!(interest.is_writable());
+    /// ```
+    ///
+    /// [`Poll`]: struct.Poll.html
+    #[inline]
+    pub fn both() -> Interests {
+        Interests(READABLE | WRITABLE)
+    }
+    
+    /// Returns true if the value includes readable readiness
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Missing a "." at the end of the sentence, same for `is_writable`.

File: src/event_imp.rs [hunk 7]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
...
+    /// ```
+    /// use mio::Interests;
+    ///
+    /// let interest = Interests::readable();
+    /// let interest_usize = interest.as_usize();
+    /// let interest2 = Interests::from_usize(interest_usize);
+    ///
+    /// assert_eq!(interest, interest2);
+    /// ```
+    pub fn from_usize(val: usize) -> Interests {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Lets not make this public. Or do you have a use case for that? Same for `as_usize`.

File: src/event_imp.rs [hunk 8]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
...
+}
+
+impl<T: Into<Interests>> ops::BitOrAssign<T> for Interests {
+    #[inline]
+    fn bitor_assign(&mut self, other: T){
+        self.0 |= other.into().0;
+    }
+}
+
+impl<T: Into<Interests>> ops::Sub<T> for Interests {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Do we want to commit to this? Are there a lot of cases where you want to remove interests?

File: src/sys/unix/dlsym.rs [hunk 9]
```diff
@@ -10,7 +10,7 @@ macro_rules! dlsym {
         static $name: ::sys::unix::dlsym::DlSym<unsafe extern fn($($t),*) -> $ret> =
             ::sys::unix::dlsym::DlSym {
                 name: concat!(stringify!($name), "\0"),
-                addr: ::std::sync::atomic::ATOMIC_USIZE_INIT,
+                addr: ::std::sync::atomic::AtomicUsize::new(0),
```
Comment (Thomasdezeeuw) [inline src/sys/unix/dlsym.rs]: Remove this, mio should still work with older versions of Rust. (I'm aware this creates a deprecation warning).

File: src/sys/unix/epoll.rs [hunk 10]
```diff
@@ -88,9 +88,9 @@ impl Selector {
     }
 
     /// Register event interests for the given IO handle with the OS
-    pub fn register(&self, fd: RawFd, token: Token, interests: Ready, opts: PollOpt) -> io::Result<()> {
+    pub fn register(&self, fd: RawFd, token: Token, interests: Interests, opts: PollOpt) -> io::Result<()> {
         let mut info = libc::epoll_event {
-            events: ioevent_to_epoll(interests, opts),
+            events: ioevent_to_epoll(Ready::from_usize(interests.as_usize()), opts),
```
Comment (Thomasdezeeuw) [inline src/sys/unix/epoll.rs]: Why not just use `Intersts` in `ioevent_to_epoll`?

File: src/sys/unix/kqueue.rs [hunk 11]
```diff
@@ -19,7 +19,7 @@ use sys::unix::io::set_cloexec;
 /// registered with the `Selector`. If a type that is previously associated with
 /// a `Selector` attempts to register itself with a different `Selector`, the
 /// operation will return with an error. This matches windows behavior.
-static NEXT_ID: AtomicUsize = ATOMIC_USIZE_INIT;
+static NEXT_ID: AtomicUsize = AtomicUsize::new(0);
```
Comment (Thomasdezeeuw) [inline src/sys/unix/kqueue.rs]: Please remove, same as with the other use of `AtomicUsize`.

File: src/sys/unix/ready.rs [hunk 12]
```diff
@@ -402,6 +402,12 @@ impl From<UnixReady> for Ready {
     }
 }
 
+impl From<UnixReady> for Interests {
```
Comment (Thomasdezeeuw) [inline src/sys/unix/ready.rs]: Why is this needed?

File: src/event_imp.rs [hunk 13]
```diff
@@ -1044,6 +1044,216 @@ fn test_debug_ready() {
...
+/// Interests are used in registering [`Evented`] handles with [`Poll`],
+/// they indicate what readiness should be monitored for. For example if a
+/// socket is registered with readable interests and the socket becomes
+/// writable, no event will be returned from [`poll`].
+///
+/// [`Poll`]: struct.Poll.html                                                 
+/// [`readable`]: #method.readable
+/// [`writable`]: #method.writable
+#[derive(Copy, PartialEq, Eq, Clone, PartialOrd, Ord)]
+pub struct Interests(usize);
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: In mio-st I've made this `NonZeroU8` to make `Option<Interests>` the same size as `Interests`. But that can come later (in a future pr).

File: src/event_imp.rs (hunk 1 above)
Comment (carllerche) [reply src/event_imp.rs]: I _think_ `rustfmt` will take care of this. I would recommend running a `rustfmt` on this branch, it should hopefully make the merge easier.

File: src/event_imp.rs (hunk 3 above)
Comment (carllerche) [reply src/event_imp.rs]:
```suggestion
    /// Returns `Interests` representing readable readiness.
```

File: src/sys/unix/dlsym.rs (hunk 9 above)
Comment (carllerche) [reply src/sys/unix/dlsym.rs]: @Thomasdezeeuw I made this change on master. Mio 0.7 can target a modern Rust :+1:

File: src/sys/unix/ready.rs (hunk 12 above)
Comment (FXTi) [reply src/sys/unix/ready.rs]: I suggest add platform-specific readiness in `Interests` to support Unix
And conversion from `Interests` to `Ready` to simplify code and get rid of `Interests::from_usize` and `Interests::to_usize`
Since `Interests` should be the subset of `Ready` and never be `empty`

File: src/sys/unix/epoll.rs (hunk 10 above)
Comment (FXTi) [reply src/sys/unix/epoll.rs]: Because this:
```rust
    pub fn push_event(&mut self, event: Event) {
        self.events.push(libc::epoll_event {
            events: ioevent_to_epoll(event.readiness(), PollOpt::empty()),
            u64: usize::from(event.token()) as u64,
        });
    }
```

`Event::readiness` return `Ready` and `ioevent_to_epoll` accept it

**Comment by Thomasdezeeuw** [comment]
@FXTi can you rebase on master again, you're currently adding a lot of code that was just removed, e.g. `src/channel.rs`.

File: src/sys/unix/epoll.rs (hunk 10 above)
Comment (carllerche) [reply src/sys/unix/epoll.rs]: Maybe we create two separate fns?

File: src/sys/unix/ready.rs (hunk 12 above)
Comment (carllerche) [reply src/sys/unix/ready.rs]: Ideally, we can avoid implementing the trait. We can add a `pub(crate)` fn for the conversion. Implementing the `From` trait makes the conversion part of the public API.

File: src/sys/unix/epoll.rs (hunk 10 above)
Comment (FXTi) [reply src/sys/unix/epoll.rs]: `ready_to_epoll` and `interests_to_epoll` ?

**Comment by FXTi** [comment]
Have checked and removed them

File: test/test_write_then_drop.rs [hunk 14]
```diff
@@ -65,9 +65,9 @@ fn write_then_deregister() {
 
     let poll = Poll::new().unwrap();
 
-    a.register(&poll, Token(1), Ready::readable(), PollOpt::edge())
+    a.register(&poll, Token(1), Interests::readable(), PollOpt::edge())
         .unwrap();
-    s.register(&poll, Token(3), Ready::empty(), PollOpt::edge())
```
Comment (Thomasdezeeuw) [inline test/test_write_then_drop.rs]: Same, input is changed.

File: test/test_write_then_drop.rs [hunk 15]
```diff
@@ -14,9 +14,9 @@ fn write_then_drop() {
...
```
Comment (Thomasdezeeuw) [inline test/test_write_then_drop.rs]: Do it matter that the input is changed here @carllerche?

File: src/event_imp.rs [hunk 16]
```diff
@@ -981,6 +986,450 @@ fn test_debug_ready() {
...
+    /// let interest = Interests::hup();
+    ///
+    /// assert!(interest.is_hup());
+    /// ```
+    ///
+    /// [`Poll`]: struct.Poll.html
+    /// [readiness]: ../struct.Poll.html#readiness-operations
+    #[inline]
+    #[cfg(unix)]
+    pub fn hup() -> Interests {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: I don't believe hup interests doesn't make sense as this is registered automatically.

File: src/event_imp.rs [hunk 17]
```diff
@@ -981,6 +986,450 @@ fn test_debug_ready() {
...
+    /// let interest = Interests::error();
+    ///
+    /// assert!(interest.is_error());
+    /// ```
+    ///
+    /// [`Poll`]: struct.Poll.html
+    /// [readiness]: ../struct.Poll.html#readiness-operations
+    #[inline]
+    #[cfg(unix)]
+    pub fn error() -> Interests {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Same as with hup, also registered by default.

File: src/event_imp.rs [hunk 18]
```diff
@@ -981,6 +986,450 @@ fn test_debug_ready() {
...
+    ///
+    /// let interest = Interests::priority();
+    ///
+    /// assert!(interest.is_priority());
+    /// ```
+    ///
+    /// [`Poll`]: struct.Poll.html
+    #[inline]
+    #[cfg(any(target_os = "linux", target_os = "android", target_os = "solaris"))]
+    pub fn priority() -> Interests {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: I believe priority is also registered by default.

File: src/event_imp.rs [hunk 19]
```diff
@@ -981,6 +986,450 @@ fn test_debug_ready() {
...
+    /// let interest = Interests::readable();
+    ///
+    /// assert_eq!(Ready::readable(),interest.to_ready());
+    /// ```
+    pub fn to_ready(&self) -> Ready {
+        Ready(self.0)
+    }
+}
+
+impl<T: Into<Interests>> ops::BitOr<T> for Interests {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Do we have anything that can be converted into `Interests`? Otherwise this can use `Interests` directly.

File: src/event_imp.rs [hunk 20]
```diff
@@ -981,6 +986,450 @@ fn test_debug_ready() {
...
+
+impl<T: Into<Interests>> ops::SubAssign<T> for Interests {
+    #[inline]
+    fn sub_assign(&mut self, other: T) {
+        self.0 &= !other.into().0;
+    }
+}
+
+impl fmt::Debug for Interests {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Depending on what interests remain this needs to be updated.

File: src/event_imp.rs [hunk 21]
```diff
@@ -981,6 +986,450 @@ fn test_debug_ready() {
...
+    ///
+    /// let interest = Interests::both();
+    ///
+    /// assert!(interest.is_readable());
+    /// assert!(interest.is_writable());
+    /// ```
+    ///
+    /// [`Poll`]: struct.Poll.html
+    #[inline]
+    pub fn both() -> Interests {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: This shouldn't be public because then `Interests` can extended in the future.

File: src/event_imp.rs [hunk 22]
```diff
@@ -981,6 +986,450 @@ fn test_debug_ready() {
...
+    /// # Examples
+    ///
+    /// ```
+    /// use mio::{Interests, Ready};
+    ///
+    /// let interest = Interests::readable();
+    ///
+    /// assert_eq!(Ready::readable(),interest.to_ready());
+    /// ```
+    pub fn to_ready(&self) -> Ready {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: This shouldn't be public and preferably not even exists and conceptually interests and ready are different.

File: test/test_echo_server.rs [hunk 23]
```diff
@@ -203,16 +214,19 @@ impl EchoClient {
             }
         };
 
-        if !self.interest.is_empty() {
+        if let Some(x) = self.interest {
+            //Interests can only be READABLE / WRITABLE.
+            /*
```
Comment (Thomasdezeeuw) [inline test/test_echo_server.rs]: I think this can be removed.

File: test/test_register_deregister.rs [hunk 24]
```diff
@@ -97,6 +101,7 @@ pub fn test_register_deregister() {
     assert!(events.iter().next().is_none());
 }
 
+/* Since Interests is used in register, there's no reason to test this.
```
Comment (Thomasdezeeuw) [inline test/test_register_deregister.rs]: This test can be removed now.

File: test/test_uds_shutdown.rs [hunk 25]
```diff
@@ -26,7 +26,8 @@ impl EchoConn {
             buf: None,
             mut_buf: Some(ByteBuf::mut_with_capacity(2048)),
             token: None,
-            interest: Ready::hup(),
+            interest: Interests::from_usize(Ready::hup()),
+            interest: Some(Interests::hup()),
```
Comment (Thomasdezeeuw) [inline test/test_uds_shutdown.rs]: Double field.

File: test/test_unix_echo_server.rs [hunk 26]
```diff
@@ -209,16 +220,19 @@ impl EchoClient {
             }
         };
 
-        if !self.interest.is_none() {
+        if let Some(x) = self.interest {
+            //Interests can only be READABLE / WRITABLE.
```
Comment (Thomasdezeeuw) [inline test/test_unix_echo_server.rs]: I think this can be removed.

**Comment by stjepang** [comment]
Just a really minor nit. I see `interest: Interests` everywhere, which should probably be changed to `interests: Interests`.

File: src/event_imp.rs (hunk 16 above)
Comment (FXTi) [reply src/event_imp.rs]: Roger

File: src/event_imp.rs (hunk 17 above)
Comment (FXTi) [reply src/event_imp.rs]: Roger

File: src/event_imp.rs (hunk 18 above)
Comment (FXTi) [reply src/event_imp.rs]: Roger

File: src/event_imp.rs (hunk 21 above)
Comment (FXTi) [reply src/event_imp.rs]: Should `Interests::both()` even exist?
Maybe `Interests::all()` is better  since  `Interests` will be extended in the future.

File: src/event_imp.rs (hunk 19 above)
Comment (FXTi) [reply src/event_imp.rs]: I will use `Interests` directly.

**Comment by FXTi** [comment]
@stjepang Have done that

File: src/event_imp.rs (hunk 22 above)
Comment (FXTi) [reply src/event_imp.rs]: I will use `pub(crate)` to restrict this in crate.

File: src/event_imp.rs (hunk 13 above)
Comment (FXTi) [reply src/event_imp.rs]: I will do that in a future pr.

File: src/sys/unix/epoll.rs (hunk 10 above)
Comment (FXTi) [reply src/sys/unix/epoll.rs]: Maybe just use `Interests::to_ready` to convert and use one `ioevent_to_epoll`?

File: test/test_write_then_drop.rs (hunk 15 above)
Comment (carllerche) [reply test/test_write_then_drop.rs]: I don't think so? I'm not 100% sure what is going on in the original test. Since `Interests` cannot be empty, this change seems fine (if tests pass).

**Comment by FXTi** [comment]
@carllerche @Thomasdezeeuw Is this ready to go? Or still change requested.

File: src/event_imp.rs (hunk 21 above)
Comment (Thomasdezeeuw) [reply src/event_imp.rs]: Lets make this private for now and we'll figure out `both` vs `all` in future pr.

File: src/event_imp.rs [hunk 27]
```diff
@@ -981,6 +986,335 @@ fn test_debug_ready() {
...
+        {
+            if self.is_lio() {
+                if one {
+                    write!(fmt, " | ")?
+                }
+                write!(fmt, "{}", "LIO")?;
+                one = true
+            }
+        }
+        if !one {
```
Comment (Thomasdezeeuw) [inline src/event_imp.rs]: Maybe make this a `debug_assert(!one, "printing empty interests")`.

**Comment by FXTi** [comment]
I've also learned something, so thank you : )
